### PR TITLE
Appsre 3576 Try to serve API errors from Cache

### PR DESCRIPTION
**Pull request by andreugallofre**
This MR aims to handle Github API errors and try to return responses from the cache as part of APPSRE-3576

### CODE REVIEW

**Comment by apahim** [comment]
I do have some issues with this proposal... serving from cache after 10s is a bad user experience. I'd rather make explicit there's a problem in the API, so we can use that information to improve the offline_mode detection thread. Also, it will count for the OFFLINE metric without really being in offline mode, so we will find ourselves in the situation where we think we are good, serving from cache, but users are only getting each response after 10s.

So, I'd first invest in the offline_mode detection thread, make it more effective. I'm certain that a good detection mechanism can mitigate the need to mask issues in the API like this.

**Comment by andreugallofre** [comment]
In my opinion, serving the request after 10s (if possible) it's not the best scenario, but It's better than waiting those 10s and getting a 502 error from the Mirror.

I agree that we might need to invest some time on re-think if we want to say that we are on offline mode after getting a timeout and serving from the cache and find better ways to detect if we should switch from online to offline. I'll write a proposal to improve this.

**Comment by apahim** [comment]
I guess it is fine to serve from cache on timeout, if we don't call it "OFFLINE_HIT", so we can tell that response apart from the offline mode. SOmething like API_ERROR_HIT, if that makes sense.

But ultimatelly, we want to make the offline detection mechanism so good that those issue won't really happen.

Just my 2 cents, don't block this work on me please.

File: ghmirror/core/mirror_requests.py
Comment (rporres) [inline ghmirror/core/mirror_requests.py:179]: If I understand correctly the code, you are only getting here for `GET` requests. Why do you need to pass `data`?

File: ghmirror/core/mirror_requests.py
Comment (rporres) [inline ghmirror/core/mirror_requests.py:61]: This is a wrapper for `GET` requests to mark them as valid or invalid. I would name it accordingly

File: ghmirror/core/mirror_requests.py
Comment (rporres) [inline ghmirror/core/mirror_requests.py:66]: I don't see this method doing all that. It just wraps the request conveniently to decide if it's valid or not.

File: ghmirror/core/mirror_requests.py
Comment (rporres) [inline ghmirror/core/mirror_requests.py:107]: Missing doc

File: ghmirror/core/mirror_requests.py
Comment (rporres) [inline ghmirror/core/mirror_requests.py:114]: why rate limited? That message should only be logged when we get a 403, isn't it?

File: ghmirror/core/mirror_requests.py
Comment (rporres) [inline ghmirror/core/mirror_requests.py:74]: I'm not convinced we should try to serve from cache on every 4xx. Most likely only in 403 and rate limited. But maybe you have already discussed this...

File: ghmirror/core/mirror_requests.py
Comment (rporres) [inline ghmirror/core/mirror_requests.py:139]: how is this related to the overall change?

Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:149]: I needed the refactor of this as `online_request` function was using the maximum number of local variables that Pylint allow, and to add the `valid_request` I did this

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:76]: That's what the code is doing, `raise_for_status()` raises an HTTPError exception if the status code of the request is 4xx or 5xx that we are capturing and trying to serve the content from the cache

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py]: The original request was passing the data, and I replicated the same in the wrapper, but maybe does not make sense and we can remove it

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py]: Forgot to change the log message as I changed the behaviour

File: ghmirror/core/mirror_requests.py
Comment (rporres) [reply ghmirror/core/mirror_requests.py:76]: the code is catching every 4xx whereas I doubt we should serve from cache a request that's returning a 404, for example.

File: ghmirror/core/mirror_requests.py
Comment (rporres) [reply ghmirror/core/mirror_requests.py]: Are you sure about that? I don't see it here: https://github.com/app-sre/github-mirror/blob/master/ghmirror/core/mirror_requests.py#L121-L125

**Comment by andreugallofre** [comment]
@rporres Changes are ready for review :)

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:85]: We will never reach to here with the empty string as the function is returning a bolean along with the header, and we only get to here if `serve_from_cache` is True
```
(serve_from_cache, response_header) = _should_error_response_be_served_from_cache(resp)
if serve_from_cache:
  XXX
```

https://github.com/app-sre/github-mirror/pull/63/files/7ffac9e2f66280bc206d0eb29b1668dc3a4ab650#diff-b65064d2dfc0b574509b9afa3a22ef4675bdbe4f2ff285de9b5ce1288c7faaecR76-R77

If it's false and the string is none, we return a cache miss directly 

https://github.com/app-sre/github-mirror/pull/63/files/7ffac9e2f66280bc206d0eb29b1668dc3a4ab650#diff-b65064d2dfc0b574509b9afa3a22ef4675bdbe4f2ff285de9b5ce1288c7faaecR88-R90

But maybe is not the best way to do it (?)

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:208]: 👍 Done!

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:218]: https://github.com/app-sre/github-mirror/pull/63#discussion_r701109929

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:194]: Moved the logic to decide if we should make a new request inside, to be able to return the response directly from the function instead of checking the output and returning in the "main", I understand  this as part of this PR, but I can split this into another PR

**Comment by andreugallofre** [comment]
@rporres @apahim Take a look at the changes / comments when able :) Thanks!

File: ghmirror/core/mirror_requests.py
Comment (rporres) [reply ghmirror/core/mirror_requests.py:194]: ok, leave it here

File: ghmirror/core/mirror_requests.py
Comment (rporres) [reply ghmirror/core/mirror_requests.py:196]: If I understand correctly the code, you still don't know if you have a cache miss as `_online_request` you are going to see if result code is 304, isn't it?

File: tests/functional/test_api.py
Comment (rporres) [inline tests/functional/test_api.py:597]: nit: missing EOL at the end of the file

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:196]: I'm adding the headers to the response that comes from the API, but if the response code is 304, we will be returning either a new request made to the api with the correct headers (Cache miss) or the response that came from the Cache with the CACHE_HIT header, and the original response is discarded. I'm trying to look for a way to clarify the code

Comment (rporres) [reply ghmirror/core/mirror_requests.py:196]: Hmm, yes, headers will be rewritten in the `304` `if` branch but you would still get an incorrect log message. That has to be the reason why tests are working.

Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:197]: Didn't saw that, 100% true, this fix should log the correct status of the request

File: ghmirror/core/mirror_requests.py
Comment (apahim) [reply ghmirror/core/mirror_requests.py:85]: So you are using an empty string for failing the "truth value test", probably because you care about types, so it might look weird not to return a string there, but python doesn't care about types. You can safelly return `None` instead of an empty string, so we know we have to do the right thing to it, otherwise the code will break.

That's just a question of how we use the language, almost a style thing, but as long as you're doing the right thing, which you are, you don't have to do as I say :)

Comment (apahim) [reply ghmirror/core/mirror_requests.py:85]: Anyway, what I'd do: `_should_error_response_be_served_from_cache` would return either a string or None. If it has a string, to use it, if it is None, you go away :)

File: ghmirror/core/mirror_requests.py
Comment (apahim) [inline ghmirror/core/mirror_requests.py:210]: it's kinda implicit that you're returning a tuple, bu you can be more explicit here:
```
    :rtype: tuple[bool, string]
```

File: ghmirror/core/mirror_requests.py
Comment (apahim) [reply ghmirror/core/mirror_requests.py:85]: Something like:
```python
    serve_from_cache_header = get_serve_from_cache_header()
    
    # early checkout :)
    if not serve_from_cache_header:
        return resp
    
    # process cached_response inside the "if" testing it
    if cached_response is not None:
        LOG.info('%s GET CACHE_HIT %s', response_header, url)
        cached_response.headers['X-Cache'] = serve_from_cache_header + '_HIT'
        return cached_response
    
    # last resort needs no extra processing
    LOG.info('%s GET CACHE_MISS %s', response_header, url)
    resp.headers['X-Cache'] = serve_from_cache_header + '_MISS'
    return resp
```

File: ghmirror/core/mirror_requests.py
Comment (rporres) [reply ghmirror/core/mirror_requests.py:197]:
```
    _cache_response(resp, cache, cache_key)
```
That should go inside the conditional as we will be serving cached data otherwise

Please add a nice comment explaining how we have arrived into that conditional since the logic is inside `_online_request`

File: ghmirror/core/mirror_requests.py
Comment (rporres) [inline ghmirror/core/mirror_requests.py:76]:
```suggestion
        error_response_header = _should_error_response_be_served_from_cache(resp)
```
This will make things clearer as to why we return if it is `None`

**Review by rporres — APPROVED** [review]
This looks good to me, I guess it is as clear as the current solution could be :)

Please squash your commits in the minimum that are needed.

Before merging, let's get the +1 from @apahim

**Comment by apahim** [comment]
Oh, I'm cool with it... Didn't verify, but it looks solid :)

**Pull request merged**
